runtime/gateway.go

```diff
@@ -624,3 +686,27 @@ func GetHostname() string {
 	}
 	return host
 }
+
+// shutdownTChannelServer gracefully shuts down the tchannel server, blocks until the shutdown is
+// complete or the timeout has reached if there is one associated with the given context
+func (gateway *Gateway) shutdownTChannelServer(ctx context.Context) error {
+	shutdownPollInterval := defaultShutdownPollInterval
+	if gateway.Config.ContainsKey("shutdown.pollInterval") {
+		shutdownPollInterval = time.Duration(gateway.Config.MustGetInt("shutdown.pollInterval")) * time.Millisecond
+	}
+	ticker := time.NewTicker(shutdownPollInterval)
+	defer ticker.Stop()
+
+	gateway.tchannelServer.Close()
```

Comment:
does tchannelServer close also shuts down without affecting active connections?

Reply:
It does, but it returns immediately, i.e., after it returns, the shutdown might have not finished.

Reply:
> what happens when we attempt to shuts down servers and send sigterm but failed to shut down server?
The errors are logged and the main thread returns.

> Will server still taking incoming requests?
It depends, if the shutdown fails to close the open listeners, then yes.

> Since the gateway.Close called only once, what happens if there's a long live hanging connections on HTTP server?
It waits indefinitely for connections to return to idle and then shut down. If the connection does not become idle before the timeout, the main thread returns.